soda/contracts/docs/01_writing_contract_yaml_files/03_writing_contract_checks.md

```diff
@@ -0,0 +1,336 @@
+* This ignoring of missing values probably doesn't work when using valid_values_column configuration
+
+### Uniqueness check
+
+Example of the simplest uniqueness check
+```yaml
+dataset: CUSTOMERS
+columns:
+- name: id
+  checks:
+  # Fail when there are duplicates
+  - type: duplicate_count
+```
+
+### Freshness check
+
+Checks if there are rows indicating recent data has been added.  It assumes there is a column that represents a timestamp like
+an event time or so.  The check looks for the maximum value in the column and verifies if that maximum value is not older than
+a given threshold time period.
+
+Example
+```yaml
+dataset: CUSTOMERS
+checks:
+    - type: freshness_in_hours
+      fail_when_greater_than: 6
+```
+
+| All freshness check types    |
+|------------------------------|
+| `type: freshness_in_days`    |
+| `type: freshness_in_hours`   |
+| `type: freshness_in_minutes` |
+
+### Avg, sum and other SQL aggregation checks
+
+Exmple of an average check
+```yaml
+dataset: CUSTOMERS
+columns:
+- name: size
+  checks:
+  # Fail when the average is not between 10 and 20
+  - type: avg
+    fail_when_not_between: [10, 20]
+```
+
+| Numeric SQL aggregation check types |
+|-------------------------------------|
+| `type: avg`                         |
+| `type: sum`                         |
+| `type: min`                         |
+| `type: max`                         |
+| `type: stddev`                      |
+| `type: stddev_pop`                  |
+| `type: stddev_samp`                 |
+| `type: variance`                    |
+| `type: var_pop`                     |
+| `type: var_samp`                    |
+
+### Multi-column uniqueness check
+
+Example of a multi columns duplicates check
+
+```yaml
+dataset: CUSTOMERS
+columns:
+- ...
+checks:
+- type: multi_column_duplicates
+  columns: ['country_code', 'zip']
+```
+
+### User defined SQL checks
+
+Example of a user-defined SQL expression check
+
+```yaml
+dataset: CUSTOMERS
+columns:
+- name: country
+  checks:
+  - type: sql_expression
+    metric: us_count
+    metric_sql_expression: COUNT(CASE WHEN country = 'US' THEN 1 END)
+    fail_when_not_between: [100, 120]
+```
+
+Example of a user-defined SQL query check
+
+```yaml
+dataset: CUSTOMERS
+columns:
+- ...
+checks:
+- type: user_defined_sql
+  metric: us_count
+  query: |
+    SELECT COUNT(*)
+    FROM {table_name}
+    WHERE country = 'US'
+  fail_when_between: [0, 5]
+```
+
+### Filter
+
+Sometimes new data is already appended in an incremental table and you only want to run the
+contract checks on the new data.  For this, a SQL filter can be used.
+
+The following check requires a variable to be passed.
+
+```yaml
+dataset: CUSTOMERS
+sql_filter: |
+  created > ${FILTER_START_TIME}
+columns:
+  - name: id
+  - name: created
+checks:
+  - type: row_count
+    fail_when_less_than: 10
+```
+
+### Thresholds
```

Comment:
How set are we on the new threshold language? Contracts/Checks do not have deep nesting, so I would propose that we evaluate an even more verbose syntax like
```
  - type: row_count
    fail:
      - not_between: [100, 120]
```
or something like that, and perhaps even introduce shortcuts like `lte`, `gt` alongside the verbose options, what do you think?

Reply:
I don't like the nesting.  The longer form assumes that most people will set up the code completion with the json schema.  Then it's easier to start with fail_| and get code completions.  Extra nesting also means more indentation errors etc.  

I m also not a fan of long and short synonyms.  Everyone should see the same.  Readabilty (and hence the long form) wins in this case i think.